fatrace.c

```diff
@@ -335,6 +339,17 @@ print_event (const struct fanotify_event_metadata *data,
     snprintf (printbuf, sizeof (printbuf), "/proc/%i", data->pid);
     int proc_fd = open (printbuf, O_RDONLY | O_DIRECTORY);
     if (proc_fd >= 0) {
+        /* get ppid */
+        if (option_ancestors) {
+            int ppid_fd = openat (proc_fd, "stat", O_RDONLY);
+            ssize_t len = read (ppid_fd, statbuf, sizeof (statbuf));
+            close (ppid_fd);
+            if (len >= 0 && sscanf(statbuf, "%*d (%*[^)]) %*c %d", &ppid) == 1)
```

Comment:
You also need to check that the `len < sizeof(statbuf)`, as otherwise it's not nul terminated.

You also have to set `statbuf[len] = '\0'` *before* calling sscanf to ensure it's nul terminated. The `static` only gives you a nulled array the first time, but afterwards if you first read a long and then a shorter file content you will get random junk at the end.

Reply:
It's ok now I think.

Reply:
I don't see `statbuf[len> = '\0'` in curent code.

Reply:
If you can explain why it's ok or open it as a separate issue, I think we can then close this PR.

Reply:
Argh, sorry -- I meant to respond to the "parent pid loop complexity" thread, then wondered where my response went, and sent it again. So *that's* where it ended up. Indeed, this should also be fixed. Thanks for spotting!